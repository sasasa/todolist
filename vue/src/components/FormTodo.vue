<template>
  <form v-on:submit.prevent="submitHandler">
    <validation-observer v-slot="{ invalid }">
      <validation-provider name="内容" rules="required|min:5" v-slot="{ errors }">
        <div class="formbox">
          <label>内容：<input autocomplete="off" type="text" name="content" v-model.trim="fm.content"></label>
          <span class="error">{{ errors[0] }}</span>
          <span class="error" v-show="!errors[0] && error_content != ''">{{ error_content }}</span>
        </div>
      </validation-provider>
      <validation-provider name="〆切" rules="required" v-slot="{ errors }">
        <div class="formbox">
          <label>〆切：<input v-bind:min="minDay" type="date" name="due_on" v-model="fm.due_on"></label>
          <span class="error">{{ errors[0] }}</span>
          <span class="error" v-show="!errors[0] && error_due_on != ''">{{ error_due_on }}</span>
        </div>
      </validation-provider>
      <div class="formbox">
        <label>終了：<input type="checkbox" name="finished" v-model="fm.finished"></label>
      </div>
      <div class="formbox">
        <input class="btn-square" v-bind:disabled="invalid || user_invalid" type="submit" v-bind:value="submitWord">
      </div>
    </validation-observer>
  </form>
</template>

<script>
import { now } from '@/helpers/date'
import { ValidationProvider, ValidationObserver } from 'vee-validate';

export default {
  name: 'FormTodo',
  props: {
    fm: {
      type: Object,
      required: true,
    },
    server_errors: {
      type: Object,
      required: false,
    },
    submitWord: {
      type: String,
      required: true,
    },
  },
  components: {
    ValidationProvider,
    ValidationObserver,
  },
  data() {
    return {
      error_content: '',
      error_due_on: '',
      remember_content: '',
      remember_error_content: '',
      user_invalid: true,
    }
  },
  watch: {
    server_errors() {
      const e = this.server_errors
      this.error_content = e.content ? "内容" + e.content[0] : ''
      this.error_due_on = e.due_on ? "〆切" + e.due_on[0] : ''
      this.remember_content = this.fm.content
      this.user_invalid = true
    },
    fm: {
      handler() {
        if (this.fm.content === this.remember_content) {
          // エラーになった際の文言を再度入力されたらエラー表示する
          this.error_content = this.remember_error_content
          this.user_invalid = true
        }
        else if (this.error_content !== '' && this.fm.content.length >= 5) {
          // 重複エラーの際、入力が変更されたらエラー文言を消す
          this.remember_error_content = this.error_content
          this.error_content = ''
          this.user_invalid = false
        }
        else if (this.fm.content.length >= 5) {
          this.user_invalid = false
        }
      },
      deep: true,
    },
  },
  computed: {
    minDay() {
      return now()
    }
  },
  methods: {
    submitHandler() {
      this.$emit('submit', { todo: Object.assign({}, this.fm) })
    },

  }
}
</script>


<style lang="scss" scoped>
.add {
  width: 100%;
  min-height: 400px;
  position: relative;
}
.error {
  color: #f00000;
  height: 1em;
  display: inline-block;
  font-size: 0.75rem;
}
.formbox {
  margin-bottom: 10px;
  width: 18em;
  text-align: right;
}
.formbox input[type='text'],
.formbox input[type='date']
{
  padding: 6px;
  width: 14em;
}
label {
  display: block;
}
</style>