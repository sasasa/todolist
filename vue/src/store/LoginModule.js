import { buildModuleTypes } from '@/helpers/store';
// import fetch from 'cross-fetch'
import 'cross-fetch/polyfill';

const URL = process.env.VUE_APP_API_URL_BASE

const moduleName = 'LoginModule';


export const namespaced = true;

export const state = {
  errors: [
  ],
  signup_errors: {
  },
  reset_password_errors: {
  },
  signup_success: false,
  send_email_success: false,
  reset_password_success: false,
  user: {
    authentication_token: '',
    password: '',
    email: '',
    id: 0,
  },
  login: false,
};

const GETTER_TYPES = {
  USER: 'USER',
  LOGIN: 'LOGIN',
  ERRORS: 'ERRORS',
  SIGNUP_ERRORS: 'SIGNUP_ERRORS',
  SIGNUP_SUCCESS: 'SIGNUP_SUCCESS',
  SEND_EMAIL_SUCCESS: 'SEND_EMAIL_SUCCESS',
  RESET_PASSWORD_SUCCESS: 'RESET_PASSWORD_SUCCESS',
  RESET_PASSWORD_ERRORS: 'RESET_PASSWORD_ERRORS',
};
const MUTATION_TYPES = {
  USER: 'USER',
  LOGIN: 'LOGIN',
  ERRORS: 'ERRORS',
  SIGNUP_ERRORS: 'SIGNUP_ERRORS',
  SIGNUP_SUCCESS: 'SIGNUP_SUCCESS',
  SEND_EMAIL_SUCCESS: 'SEND_EMAIL_SUCCESS',
  RESET_PASSWORD_SUCCESS: 'RESET_PASSWORD_SUCCESS',
  RESET_PASSWORD_ERRORS: 'RESET_PASSWORD_ERRORS',
};
const ACTION_TYPES = {
  LOGIN_ASYNC: 'LOGIN_ASYNC',
  SIGNUP_ASYNC: 'SIGNUP_ASYNC',
  LOGOUT_ASYNC: 'LOGOUT_ASYNC',
  SEND_EMAIL_ASYNC: 'SEND_EMAIL_ASYNC',
  RESET_PASSWORD_ASYNC: 'RESET_PASSWORD_ASYNC',
};

export const LOGIN_GETTER_TYPES = buildModuleTypes({
  moduleName,
  types: GETTER_TYPES,
});
export const LOGIN_MUTATION_TYPES = buildModuleTypes({
  moduleName,
  types: MUTATION_TYPES,
});
export const LOGIN_ACTION_TYPES = buildModuleTypes({
  moduleName,
  types: ACTION_TYPES,
});

export const getters = {
  [GETTER_TYPES.USER](state) {
    return state.user
  },
  [GETTER_TYPES.LOGIN](state) {
    return state.login
  },
  [GETTER_TYPES.ERRORS](state) {
    return state.errors;
  },
  [GETTER_TYPES.SIGNUP_ERRORS](state) {
    return state.signup_errors;
  },
  [GETTER_TYPES.SIGNUP_SUCCESS](state) {
    return state.signup_success;
  },
  [GETTER_TYPES.SEND_EMAIL_SUCCESS](state) {
    return state.send_email_success
  },
  [GETTER_TYPES.RESET_PASSWORD_SUCCESS](state) {
    return state.reset_password_success
  },
  [GETTER_TYPES.RESET_PASSWORD_ERRORS](state) {
    return state.reset_password_errors
  },
};

export const mutations = {
  [MUTATION_TYPES.USER](state, { user }) {
    state.user = user
  },
  [MUTATION_TYPES.LOGIN](state, { login }) {
    state.login = login
  },
  [MUTATION_TYPES.ERRORS](state, { errors }) {
    state.errors = errors
  },
  [MUTATION_TYPES.SIGNUP_ERRORS](state, { errors }) {
    state.signup_errors = errors
  },
  [MUTATION_TYPES.SIGNUP_SUCCESS](state, { success }) {
    state.signup_success = success
  },
  [MUTATION_TYPES.SEND_EMAIL_SUCCESS](state, { success }) {
    state.send_email_success = success
  },
  [MUTATION_TYPES.RESET_PASSWORD_SUCCESS](state, { success }) {
    state.reset_password_success = success
  },
  [MUTATION_TYPES.RESET_PASSWORD_ERRORS](state, { errors }) {
    state.reset_password_errors = errors
  },
};

export const actions = {
  [ACTION_TYPES.RESET_PASSWORD_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: false })

    fetch(`${URL}/users/reset_password.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'X-CSRF-Token': localStorage.csrf,
        'Content-type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.success) {
        context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: true })
      } else {
        context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: false })
        context.commit(MUTATION_TYPES.RESET_PASSWORD_ERRORS, { errors: data.errors })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error)
      alert(error)
      context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: false })
      context.commit('toggle_loading', payload, { root: true })
    })
  },
  [ACTION_TYPES.SEND_EMAIL_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: false })

    fetch(`${URL}/users/send_email.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'X-CSRF-Token': localStorage.csrf,
        'Content-type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.success) {
        context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: true })
      } else {
        context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: false })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error)
      alert(error)
      context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: false })
      context.commit('toggle_loading', payload, { root: true })
    })
  },
  [ACTION_TYPES.LOGOUT_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })

    fetch(`${URL}/users/logout.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'X-CSRF-Token': localStorage.csrf,
        'Content-type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.success) {
        context.commit(MUTATION_TYPES.USER, { user: {} })
        context.commit(MUTATION_TYPES.LOGIN, { login: false })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },


  [ACTION_TYPES.LOGIN_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })

    fetch(`${URL}/users/login.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 'Content-type': 'application/json' },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.errors) {
        const errors = data.errors
        context.commit(MUTATION_TYPES.ERRORS, { errors })
      } else {
        // LocalStorage に CSRF トークンを保存する
        localStorage.csrf = data.csrf;
  
        const user = data.user
        context.commit(MUTATION_TYPES.ERRORS, { errors: [] })
        context.commit(MUTATION_TYPES.USER, { user })
        context.commit(MUTATION_TYPES.LOGIN, { login: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.SIGNUP_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })

    fetch(`${URL}/users.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 'Content-type': 'application/json' },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      // alert(JSON.stringify(data))
      if (data.errors) {
        context.commit(MUTATION_TYPES.SIGNUP_SUCCESS, { success: false })
        context.commit(MUTATION_TYPES.SIGNUP_ERRORS, { errors: data.errors })
      } else {
        context.commit(MUTATION_TYPES.SIGNUP_SUCCESS, { success: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },
};