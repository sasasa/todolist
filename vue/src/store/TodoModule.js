import { buildModuleTypes } from '@/helpers/store';
// import fetch from 'cross-fetch'
import 'cross-fetch/polyfill';

const URL = process.env.VUE_APP_API_URL_BASE

const moduleName = 'TodoModule';


export const namespaced = true;

export const state = {
  todo: {
    content: '',
    due_on: '',
    finished: false,
    id: 0,
  },
  todos: [
  ],
  add_errors: {
    content: [],
  },
  add_success: false,
  update_errors: {
    content: [],
  },
  update_success: false,
};

const GETTER_TYPES = {
  TODOS: 'TODOS',
  TODO: 'TODO',
  ADD_ERRORS: 'ADD_ERRORS',
  ADD_SUCCESS: 'ADD_SUCCESS',
  UPDATE_ERRORS: 'UPDATE_ERRORS',
  UPDATE_SUCCESS: 'UPDATE_SUCCESS',
};
const MUTATION_TYPES = {
  GET_TODO: 'GET_TODO',
  UPDATE_TODO: 'UPDATE_TODO',
  GET_TODOS: 'GET_TODOS',
  REMOVE_TODO: 'REMOVE_TODO',
  ADD_TODO: 'ADD_TODO',
  ADD_ERRORS: 'ADD_ERRORS',
  ADD_SUCCESS: 'ADD_SUCCESS',
  UPDATE_ERRORS: 'UPDATE_ERRORS',
  UPDATE_SUCCESS: 'UPDATE_SUCCESS',
};
const ACTION_TYPES = {
  GET_TODO_ASYNC: 'GET_TODO_ASYNC',
  UPDATE_TODO_ASYNC: 'UPDATE_TODO_ASYNC',
  GET_TODOS_ASYNC: 'GET_TODOS_ASYNC',
  REMOVE_TODO_ASYNC: 'REMOVE_TODO_ASYNC',
  ADD_TODO_ASYNC: 'ADD_TODO_ASYNC',
};

export const TODO_GETTER_TYPES = buildModuleTypes({
  moduleName,
  types: GETTER_TYPES,
});
export const TODO_MUTATION_TYPES = buildModuleTypes({
  moduleName,
  types: MUTATION_TYPES,
});
export const TODO_ACTION_TYPES = buildModuleTypes({
  moduleName,
  types: ACTION_TYPES,
});

export const getters = {
  [GETTER_TYPES.TODOS](state) {
    return state.todos;
  },
  [GETTER_TYPES.TODO](state) {
    return state.todo;
  },
  [GETTER_TYPES.ADD_ERRORS](state) {
    return state.add_errors;
  },
  [GETTER_TYPES.ADD_SUCCESS](state) {
    return state.add_success;
  },
  [GETTER_TYPES.UPDATE_ERRORS](state) {
    return state.update_errors;
  },
  [GETTER_TYPES.UPDATE_SUCCESS](state) {
    return state.update_success;
  },
};

export const mutations = {
  [MUTATION_TYPES.GET_TODO](state, { todo }) {
    state.todo = todo
  },
  [MUTATION_TYPES.GET_TODOS](state, { todos }) {
    state.todos = todos
  },
  [MUTATION_TYPES.REMOVE_TODO](state, { todo }) {
    state.todos = state.todos.filter((t)=>{
      return t.id !== todo.id
    })
  },
  [MUTATION_TYPES.UPDATE_TODO](state, { todo }) {
    state.todos.forEach((t)=>{
      if(t.id === todo.id) {
        t.finished = todo.finished
        t.content = todo.content
        t.due_on = todo.due_on
      }
    })
  },
  [MUTATION_TYPES.ADD_TODO](state, { todo }) {
    state.todos.push(todo)
  },
  [MUTATION_TYPES.ADD_ERRORS](state, { errors }) {
    state.add_errors = errors
  },
  [MUTATION_TYPES.ADD_SUCCESS](state, { success }) {
    state.add_success = success
  },
  [MUTATION_TYPES.UPDATE_ERRORS](state, { errors }) {
    state.update_errors = errors
  },
  [MUTATION_TYPES.UPDATE_SUCCESS](state, { success }) {
    state.update_success = success
  },
};

export const actions = {
  [ACTION_TYPES.GET_TODO_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    fetch(`${URL}/todos/${payload.id}.json?user_email=${payload.user.email}`, { 
      method: 'GET',
      headers: {
        Authorization: `Bearer ${payload.user.authentication_token}`,
      }
    })
    .then(response => {
      if (200 <= response.status < 300 && response.ok) {
        return response.json();
      } else {
        throw new Error(`status NG ${response.status}:${response.statusText}`);
      }
    })
    .then(todo => {
      context.commit(MUTATION_TYPES.GET_TODO, { todo })
      // console.log(`--------------${ todo.id }`)
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.GET_TODOS_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    fetch(`${URL}/todos.json?user_email=${payload.user.email}`, { 
      method: 'GET',
      headers: {
        Authorization: `Bearer ${payload.user.authentication_token}`,
      }
    })
    .then(response => {
      if (200 <= response.status < 300 && response.ok) {
        return response.json();
      } else {
        throw new Error(`status NG ${response.status}:${response.statusText}`);
      }
    })
    .then(todos => {
      context.commit(MUTATION_TYPES.GET_TODOS, { todos })
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.REMOVE_TODO_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    fetch(`${URL}/todos/${payload.todo.id}.json?user_email=${payload.user.email}`, {
      method: 'DELETE',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'Content-type': 'application/json',
        'X-CSRF-Token': localStorage.csrf,
        Authorization: `Bearer ${payload.user.authentication_token}`,
      },
    })
    .then(response => {
      if (200 <= response.status < 300 && response.ok) {
        return response.json();
      } else {
        throw new Error(`status NG ${response.status}:${response.statusText}`);
      }
    })
    .then((todo) => {
      todo
      context.commit(MUTATION_TYPES.REMOVE_TODO, payload)
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.UPDATE_TODO_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.UPDATE_SUCCESS, { success: false })
    const id = payload.todo.id
    delete payload.todo.id
    fetch(`${URL}/todos/${id}.json?user_email=${payload.user.email}`, {
      method: 'PUT',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'Content-type': 'application/json',
        'X-CSRF-Token': localStorage.csrf,
        Authorization: `Bearer ${payload.user.authentication_token}`,
      },
      body: JSON.stringify(payload.todo),
    })
    .then(res => res.json())
    .then(todo => {
      if (todo.errors) {
        context.commit(MUTATION_TYPES.UPDATE_ERRORS, { errors: todo.errors })
      } else {
        context.commit(MUTATION_TYPES.UPDATE_TODO, { todo })
        context.commit(MUTATION_TYPES.UPDATE_SUCCESS, { success: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.ADD_TODO_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.ADD_SUCCESS, { success: false })
    fetch(`${URL}/todos.json?user_email=${payload.user.email}`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'Content-type': 'application/json',
        'X-CSRF-Token': localStorage.csrf,
        Authorization: `Bearer ${payload.user.authentication_token}`,
      },
      body: JSON.stringify(payload.todo),
    })
    .then(response => {
      if (200 <= response.status < 300 && response.ok) {
        return response.json();
      } else if (response.status === 422) {
        return response.json();//validation error
      } else {
        throw new Error(`status NG ${response.status}:${response.statusText}`);
      }
    })
    .then(todo => {
      if (todo.errors) {
        context.commit(MUTATION_TYPES.ADD_ERRORS, { errors: todo.errors })
      } else {
        context.commit(MUTATION_TYPES.ADD_TODO, { todo })
        context.commit(MUTATION_TYPES.ADD_SUCCESS, { success: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },
};