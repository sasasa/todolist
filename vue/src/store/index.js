

import Vue from 'vue';
import Vuex from 'vuex';
import createPersistedState from "vuex-persistedstate";

import * as TodoModule from './TodoModule';
import * as LoginModule from './LoginModule';

Vue.use(Vuex);

export default new Vuex.Store({
  strict: true,
  modules: {
    TodoModule,
    LoginModule,
  },
  state: {
    loading: false,
  },
  getters: {
    ['loading'](state) {
      return state.loading;
    },
  },
  mutations: {
    ['toggle_loading'](state) {
      state.loading = !state.loading 
    },
  },
  plugins: [createPersistedState()],
});

