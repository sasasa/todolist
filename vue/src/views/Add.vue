<template>
  <div class="add">
    <h1>Add</h1>
    <div>
      <form-todo
        v-on:submit="onSubmit"
        v-bind:fm="fm"
        v-bind:server_errors="errors"
        submitWord="追加する"
      />
    </div>
  </div>
</template>

<script>
import FormTodo from '@/components/FormTodo'
import { oneWeekLater } from '@/helpers/date'
import { TODO_GETTER_TYPES, TODO_ACTION_TYPES } from '@/store/TodoModule'
import { LOGIN_GETTER_TYPES } from '@/store/LoginModule';
import { Titleable } from '@/mixin/titleable'

export default {
  name: 'add',
  mixins: [
    Titleable,
  ],
  props: {
  },
  components: {
    FormTodo,
  },
  data() {
    return {
      title: 'todolist | 追加',

      fm: {
        content: '',
        due_on: oneWeekLater(),
        finished: false,
      }
    }
  },
  watch: {
    success(value) {
      if (value === true) {
        this.$router.push('/todos')
      }
    },
  },
  computed: {
    errors() {
      return this.$store.getters[TODO_GETTER_TYPES.ADD_ERRORS]
    },
    success() {
      return this.$store.getters[TODO_GETTER_TYPES.ADD_SUCCESS]
    },
    user() {
      return this.$store.getters[LOGIN_GETTER_TYPES.USER]
    },
  },
  methods: {
    onSubmit({ todo }) {
      this.$store.dispatch(TODO_ACTION_TYPES.ADD_TODO_ASYNC, { todo, user: this.user })
    },
  }
}
</script>


<style lang="scss" scoped>
</style>