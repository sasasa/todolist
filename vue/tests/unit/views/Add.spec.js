import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Vuex from 'vuex'
import Add from '@/views/Add'
import * as TodoModule from '../store/TodoModuleMock'
import * as LoginModule from '../store/LoginModuleMock'
import FormTodo from '@/components/FormTodo'


const localVue = createLocalVue()
// localVue.use(VueRouter)
localVue.use(Vuex)

const router = new VueRouter()

describe('Add view', () => {
  let store
  let $router
  let $store
  let todo

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        TodoModule,
        LoginModule,
      }
    })
    $router = {
      push: jest.fn()
    }
    todo = {
      id: 1,
      content: 'content test',
      due_on: '2019-10-5',
      finished: false,
    }
  })

  it('mount', () => {
    const wrapper = shallowMount(Add, {
      store,
      router,
      localVue,
    })
    expect(wrapper.isVueInstance()).toBe(true)
  })

  describe('methods', () => {
    it('onSubmit', () => {
      const wrapper = shallowMount(Add, {
        store,
        router,
        localVue,
        mocks: {
          $router,
        }
      })
      const user = LoginModule.getters[LoginModule.GETTER_TYPES.USER](LoginModule.state)

      wrapper.vm.onSubmit({ todo })
      expect.assertions(2)
      expect(TodoModule.actions[TodoModule.ACTION_TYPES.ADD_TODO_ASYNC].mock.calls).toHaveLength(1)
      expect(TodoModule.actions[TodoModule.ACTION_TYPES.ADD_TODO_ASYNC].mock.calls[0][1]).toEqual({ todo, user })
    })
  })

  describe('template', () => {
    it('@submit="onSubmit', () => {
      const mock = jest.fn()
      const wrapper = shallowMount(Add, {
        store,
        router,
        localVue,
        mocks: {
          $router,
        }
      })
      wrapper.setMethods({
        onSubmit: mock
      })
      const formTodo = wrapper.find(FormTodo)
      formTodo.vm.$emit('submit')
      expect(mock).toHaveBeenCalled()
    })

    it('snapshot', () => {
      const wrapper = shallowMount(Add, {
        store,
        router,
        localVue,
        mocks: {
          $router,
          $store,
        }
      })
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })
})