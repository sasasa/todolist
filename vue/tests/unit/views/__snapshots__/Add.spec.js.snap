// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add view template snapshot 1`] = `
<div
  class="add"
>
  <h1>
    Add
  </h1>
   
  <div>
    <form-todo-stub
      fm="[object Object]"
      submitword="追加する"
    />
  </div>
</div>
`;
